# vim: syntax=python expandtab
# Publications/citations to tools used on StaG-mwc.
publications = {
    "BBMap": "Bushnell, B. (2016). BBMap short read aligner. University of California, Berkeley, California. Available online at: http://sourceforge.net/projects/bbmap.",
    "Bowtie2": "Langmead B, Salzberg S (2012). Fast gapped-read alignment with Bowtie 2. Nature Methods. 2012, 9:357-359. https://doi.org/10.1038/nmeth.1923",
    "Bracken": "Lu J, Breitwieser FP, Thielen P, Salzberg SL (2017). Bracken: estimating species abundance in metagenomics data. PeerJ Computer Science 3:e104. https://doi.org/10.7717/peerj-cs.104",
    "CONCOCT": "Alneberg, J., Bjarnason, B.S., de Bruijn, I., Schirmer, M., Quick, J., Ijaz, U.Z., Lahti, L., Loman, N.J, Andersson, A.F, Quince, C. (2014). Binning metagenomic contigs by coverage and composition. Nature Methods, https://doi.org/10.1038/nmeth.3103",
    "GROOT": "Rowe WPM, Winn MD (2018). Indexed variation graphs for efficient and accurate resistome profiling. Bioinformatics, Volume 34, Issue 21. https://doi.org/10.1093/bioinformatics/bty387",
    "AMRPlusPlus2": "Doster E, Lakin S, Dean C, Wlfe C, Young J, Boucher C et al. (2019). MEGAREs 2.0: a database for classification of antimicrobial drug, biocide and metal resistance determinants in metagenomic sequence data. Nucleic Acids Research, Volume 48, Issue D1, 08 January 2020, Pages D561–D569, https://doi.org/10.1093/nar/gkz1010",
    "HUMAnN": "Franzosa EA*, McIver LJ*, et al. (2018). Species-level functional profiling of metagenomes and metatranscriptomes. Nat Methods 15: 962-968. https://doi.org/10.1038/s41592-018-0176-y",
    "Kaiju": "Menzel, P., Ng, K. L., & Krogh, A. (2016). Fast and sensitive taxonomic classification for metagenomics with Kaiju. Nature communications, 7, 11257. Available online at: https://github.com/bioinformatics-centre/kaiju",
    "Kraken2": "Wood, D.E., Lu, J., & Langmead, B. (2019). Improved metagenomic analysis with Kraken 2. Genome biology, 20, 257. https://doi.org/10.1186/s13059-019-1891-0",
    "Krona": "Ondov BD, Bergman NH, and Phillippy AM. Interactive metagenomic visualization in a Web browser. BMC Bioinformatics. 2011 Sep 30; 12(1):385. https://doi.org/10.1186/1471-2105-12-385",
    "MEGAHIT": "Li, D., Liu, C-M., Luo, R., Sadakane, K., and Lam, T-W., (2015). MEGAHIT: An ultra-fast single-node solution for large and complex metagenomics assembly via succinct de Bruijn graph. Bioinformatics, https://doi.org/10.1093/bioinformatics/btv033 [PMID: 25609793].",
    "MaxBin2": "Yu-Wei Wu, Blake A. Simmons, Steven W. Singer (2016). MaxBin 2.0: an automated binning algorithm to recover genomes from multiple metagenomic datasets. Bioinformatics, Volume 32, Issue 4, 15 February 2016, Pages 605–607, https://doi.org/10.1093/bioinformatics/btv638",
    "MetaBAT2": "Kang DD, Li F, Kirton E, Thomas A, Egan R, An H, Wang Z. (2019). MetaBAT 2: an adaptive binning algorithm for robust and efficient genome reconstruction from metagenome assemblies. PeerJ 7:e7359. https://doi.org/10.7717/peerj.7359",
    "MetaPhlAn": "Truong, D. T., et al. (2015). MetaPhlAn2 for enhanced metagenomic taxonomic profiling. Nature methods, 12, 902. https://doi.org/10.1038/nmeth.3589",
    "MetaWRAP": "Uritskiy, Gherman V., DiRuggiero, Jocelyne, Taylor, James (2018). MetaWRAP—a flexible pipeline for genome-resolved metagenomic data analysis. Microbiome 158:8, https://doi.org/10.1186/s40168-018-0541-1",
    "MultiQC": "Philip Ewels, Måns Magnusson, Sverker Lundin and Max Käller (2018). MultiQC: Summarize analysis results for multiple tools and samples in a single report. Bioinformatics. https://doi.org/10.1093/bioinformatics/btw354",
    "Snakemake": "Köster, Johannes and Rahmann, Sven (2012). Snakemake - A scalable bioinformatics workflow engine. Bioinformatics, Volume 28, Issue 19. https://doi.org/10.1093/bioinformatics/bts480",
    "StaG": "Boulund et al. (2018). StaG-mwc: metagenomic workflow collaboration. https://doi.org/10.5281/zenodo.1483891",
    "fastp": "Shifu Chen, Yanqing Zhou, Yaru Chen, Jia Gu (2018). fastp: an ultra-fast all-in-one FASTQ preprocessor. Bioinformatics, Volume 34, Issue 17, 01 September 2018, Pages i884–i890. https://doi.org/10.1093/bioinformatics/bty560",
    "featureCount": "Yang Liao, Gordon K. Smyth, Wei Shi (2014). featureCounts: an efficient general purpose program for assigning sequence reads to genomic features, Bioinformatics, Volume 30, Issue 7, 1 April 2014, Pages 923–930, https://doi.org/10.1093/bioinformatics/btt656",
}
